Course Human-Centered Data Science ([HCDS](https://www.mi.fu-berlin.de/en/inf/groups/hcc/teaching/winter_term_2020_21/course_human_centered_data_science.html)) - Winter Term 2020/21 - [HCC](https://www.mi.fu-berlin.de/en/inf/groups/hcc/index.html) | [Freie Universität Berlin](https://www.fu-berlin.de/)

***

# A4 - Transparency
Please use the follwing structure as a starting point. Extend and change the notebook according to your needs. This structure should help you to guide you through your analysis. This notebook is the foundation for condensing your results and writing your reflection in the end. So please read what we expect from you regarding the reflection first to structure your analysis accordingly.

In [97]:
import pandas as pd
import requests
import json
from tabulate import tabulate

In [98]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

## [1] General understanding
> What is the model about and who is using it?

* What is your model about?

`reverted`: The model used for this endpoitnt predicts whether an edit will eventually be reverted.


* Why is this model useful?
 * It is useful for quality control tools
 * Helps reviewers to find potentially damaging contributions -> make the work of filtering through the Special:RecentChanges feed easier
 * Can be used for detection and removal of damaging contributions
 * There's also the need to identify good-faith contributors



* Who is using this model? 
  * **User** The model is designed to help human editors perform critical wiki-work and to increase their productivity by automating tasks like detecting vandalism and removing edits made in bad fait.
  * **Developers** The model aims to provide data for developers of tools for wikipedia (See below)
  * (**Scientists**)


* What are stakeholder or users of ORES? 
  * Volunteer tool developers and product developers at the **Wikimedia Foundation** and **Wikimedia Deutschland**
  * Authors/Editors of articles (because their edits get assessed)
  * Editors/Reviewers


* Why is this model useful to wikipedia?

It aims to improve the quality of articles and to reduce the work of reviewers/other editors.


* What applications/projects/... within wikipedia are using this model?

3rd party tools used:
  * [Edit Review Improvements (ERI)](https://www.mediawiki.org/wiki/Edit_Review_Improvements/New_filters_for_edit_review)
  * [Huggle](https://en.wikipedia.org/wiki/Wikipedia:Huggle)


## [2] API
> What does the ORES API (v3) tell you about a specific model? What functions does the API offer?

Use the API to investigate your model: https://ores.wikimedia.org/v3/#/. What do the follwing API calls do and what do they tell you about your model?

* `https://ores.wikimedia.org/v3/scores/`
* `https://ores.wikimedia.org/v3/scores/?model_info`
* `https://ores.wikimedia.org/v3/scores/enwiki`
* `https://ores.wikimedia.org/v3/scores/enwiki?models=YOURMODELNAME&model_info`
* `https://ores.wikimedia.org/v3/scores/enwiki?models=YOURMODELNAME&revids=SOMEIDHERE`
* `https://ores.wikimedia.org/v3/scores/enwiki/REVID/YOURMODELNAME?model_info`
* `https://ores.wikimedia.org/v3/scores/enwiki/REVID/YOURMODELNAME?features=true`

> The API call `https://ores.wikimedia.org/v3/scores/` does provide information about the projects and the available models and versions for each project. 
_______________________________________________________

In [99]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/')
v3_scores = api_call.json()

In [100]:
accuracy_dict = {'project': [],'model':[],'version':[]}
v3_scores_df = pd.DataFrame(accuracy_dict, columns = ['model','project', 'version' ])

In [101]:
print('There are ',len(v3_scores_df.keys()),' Wikipedia projects.')
v3_scores.keys()

There are  3  Wikipedia projects.


dict_keys(['arwiki', 'bnwiki', 'bswiki', 'cawiki', 'cswiki', 'dewiki', 'elwiki', 'enwiki', 'enwiktionary', 'eswiki', 'eswikibooks', 'eswikiquote', 'etwiki', 'euwiki', 'fakewiki', 'fawiki', 'fiwiki', 'frwiki', 'frwikisource', 'glwiki', 'hewiki', 'hrwiki', 'huwiki', 'idwiki', 'iswiki', 'itwiki', 'jawiki', 'kowiki', 'lvwiki', 'nlwiki', 'nowiki', 'plwiki', 'ptwiki', 'rowiki', 'ruwiki', 'simplewiki', 'sqwiki', 'srwiki', 'svwiki', 'tawiki', 'testwiki', 'trwiki', 'ukwiki', 'viwiki', 'wikidatawiki', 'zhwiki'])

In [102]:
for project in v3_scores.keys():
    for model in v3_scores[project]['models'].keys():
        version = v3_scores[project]['models'][model]['version']
        v3_scores_df = v3_scores_df.append({ 'model':model, 'project': project, 'version':version}, ignore_index=True)

In [103]:
#print('ORES API (v3) provides ',len(pd.DataFrame(pd.unique(v3_scores_df['model']))),' different models')
#pd.DataFrame(pd.unique(v3_scores_df['model']))

print('The reverted model is available in ',len(v3_scores_df[v3_scores_df['model'] == 'reverted']),' projects. Mainly in version 0.5.0')
v3_scores_df[v3_scores_df['model'] == 'reverted']

The reverted model is available in  10  projects. Mainly in version 0.5.0


model       project version
3    reverted        bnwiki   0.5.0
13   reverted        elwiki   0.5.0
21   reverted  enwiktionary   0.5.0
46   reverted        glwiki   0.5.0
49   reverted        hrwiki   0.5.0
52   reverted        idwiki   0.5.0
53   reverted        iswiki   0.5.0
91   reverted        tawiki   0.5.0
98   reverted      testwiki   0.0.3
108  reverted        viwiki   0.5.0

In [104]:
for project in v3_scores.keys():
    for model in v3_scores[project]['models'].keys():
        version = v3_scores[project]['models'][model]['version']
        v3_scores_df = v3_scores_df.append({ 'model':model, 'project': project, 'version':version}, ignore_index=True)

In [105]:
print('The reverted model is available in ',len(v3_scores_df[v3_scores_df['model'] == 'reverted']),' projects. Mainly in version 0.5.0')
v3_scores_df[v3_scores_df['model'] == 'reverted']

The reverted model is available in  20  projects. Mainly in version 0.5.0


model       project version
3    reverted        bnwiki   0.5.0
13   reverted        elwiki   0.5.0
21   reverted  enwiktionary   0.5.0
46   reverted        glwiki   0.5.0
49   reverted        hrwiki   0.5.0
52   reverted        idwiki   0.5.0
53   reverted        iswiki   0.5.0
91   reverted        tawiki   0.5.0
98   reverted      testwiki   0.0.3
108  reverted        viwiki   0.5.0
117  reverted        bnwiki   0.5.0
127  reverted        elwiki   0.5.0
135  reverted  enwiktionary   0.5.0
160  reverted        glwiki   0.5.0
163  reverted        hrwiki   0.5.0
166  reverted        idwiki   0.5.0
167  reverted        iswiki   0.5.0
205  reverted        tawiki   0.5.0
212  reverted      testwiki   0.0.3
222  reverted        viwiki   0.5.0

> The API call `https://ores.wikimedia.org/v3/scores/?model_info` provides information for all models. It provides information about the following model properties of a model: `environment`, `params`, `score_schema`, `statistics`, `type` and `version`. Furthermore with this call it is possible to see which model is available for a wikipedia project of interest. 

________________________________________________

In [106]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/?model_info')
v3_scores_model_info = api_call.json()

In [107]:
print('This API call provides information on which models are available for a project of interest.')
print('For the wikipedia project hrwiki only the \"reverted\" model seems to be available.')
pd.DataFrame(v3_scores_model_info['hrwiki']['models'].keys())

This API call provides information on which models are available for a project of interest.
For the wikipedia project hrwiki only the "reverted" model seems to be available.


0
0  reverted

In [108]:
print("The modelinfo API call provides information on following properties of all models:")
print(pd.DataFrame(v3_scores_model_info['hrwiki']['models']['reverted'].keys()))

The modelinfo API call provides information on following properties of all models:
              0
0   environment
1        params
2  score_schema
3    statistics
4          type
5       version


In [109]:
def model_overview(project, model):
    env        = pd.DataFrame(v3_scores_model_info[project]['models'][model]['environment'].keys(),columns=['environment'])
    params     = pd.DataFrame(v3_scores_model_info[project]['models'][model]['params'].keys(),columns=['params'] )
    schema     = pd.DataFrame(v3_scores_model_info[project]['models'][model]['score_schema'].keys(), columns=['score_schema'])
    stats     = pd.DataFrame(v3_scores_model_info[project]['models'][model]['statistics'].keys(),columns=['statistics'] )   
    model_overview = params.join([env, stats, schema], how='outer').fillna('')
    return model_overview

In [110]:
model_overview = model_overview('hrwiki', 'reverted')

In [111]:
print('Following is an overview of the information each properties provides on an model: ')
model_overview

Following is an overview of the information each properties provides on an model: 


params            environment   statistics score_schema
0                  ccp_alpha                machine          !f1   properties
1                     center               platform   !precision        title
2                  criterion              processor      !recall         type
3                       init          python_branch     accuracy             
4              label_weights           python_build       counts             
5                     labels        python_compiler           f1             
6              learning_rate  python_implementation  filter_rate             
7                       loss        python_revision          fpr             
8                  max_depth         python_version   match_rate             
9               max_features                release       pr_auc             
10            max_leaf_nodes     revscoring_version    precision             
11     min_impurity_decrease                 system        rates             
12        min_impurity_split                version       recall             
13          min_samples_leaf                             roc_auc             
14         min_samples_split                                                 
15  min_weight_fraction_leaf                                                 
16                multilabel                                                 
17              n_estimators                                                 
18          n_iter_no_change                                                 
19          population_rates                                                 
20                   presort                                                 
21              random_state                                                 
22                     scale                                                 
23                 subsample                                                 
24                       tol                                                 
25       validation_fraction                                                 
26                   verbose                                                 
27                warm_start

In [112]:
print('The score schema of the model \"reverted\" in the project \"hrwiki\":')
v3_scores_model_info['hrwiki']['models']['reverted']['score_schema']

The score schema of the model "reverted" in the project "hrwiki":


{'properties': {'prediction': {'description': 'The most likely label predicted by the estimator',
   'type': 'boolean'},
  'probability': {'description': 'A mapping of probabilities onto each of the potential output labels',
   'properties': {'false': {'type': 'number'}, 'true': {'type': 'number'}},
   'type': 'object'}},
 'title': 'Scikit learn-based classifier score with probability',
 'type': 'object'}

In [113]:
stats_dict = {'metrics': [],'value':[]}
v3_stats_df = pd.DataFrame(stats_dict, columns = ['metrics','value' ])

for stat in v3_scores_model_info[project]['models'][model]['statistics'].keys():
    for label in v3_scores_model_info[project]['models'][model]['statistics'][stat].keys():
        v3_stats_df = v3_stats_df.append({ 'metrics':stat+' ('+label+')', 
                                          'value': v3_scores_model_info[project]['models'][model]['statistics'][stat][label]}, 
                                         ignore_index=True)

In [114]:
print('Here is an overview about the statistics of the model \"reverted\" for the wikipedia project \"hrwiki\":')
v3_stats_df

Here is an overview about the statistics of the model "reverted" for the wikipedia project "hrwiki":


metrics  \
0           !f1 (labels)   
1            !f1 (macro)   
2            !f1 (micro)   
3    !precision (labels)   
4     !precision (macro)   
5     !precision (micro)   
6       !recall (labels)   
7        !recall (macro)   
8        !recall (micro)   
9      accuracy (labels)   
10      accuracy (macro)   
11      accuracy (micro)   
12       counts (labels)   
13            counts (n)   
14  counts (predictions)   
15           f1 (labels)   
16            f1 (macro)   
17            f1 (micro)   
18  filter_rate (labels)   
19   filter_rate (macro)   
20   filter_rate (micro)   
21          fpr (labels)   
22           fpr (macro)   
23           fpr (micro)   
24   match_rate (labels)   
25    match_rate (macro)   
26    match_rate (micro)   
27       pr_auc (labels)   
28        pr_auc (macro)   
29        pr_auc (micro)   
30    precision (labels)   
31     precision (macro)   
32     precision (micro)   
33    rates (population)   
34        rates (sample)   
35       recall (labels)   
36        recall (macro)   
37        recall (micro)   
38      roc_auc (labels)   
39       roc_auc (macro)   
40       roc_auc (micro)   

                                                                           value  
0                                                {'false': 0.985, 'true': 0.375}  
1                                                                           0.68  
2                                                                          0.394  
3                                                {'false': 0.977, 'true': 0.593}  
4                                                                          0.785  
5                                                                          0.605  
6                                                {'false': 0.994, 'true': 0.274}  
7                                                                          0.634  
8                                                                          0.297  
9                                                {'false': 0.971, 'true': 0.971}  
10                                                                         0.971  
11                                                                         0.971  
12                                                 {'false': 263, 'true': 68528}  
13                                                                         68791  
14  {'false': {'false': 72, 'true': 191}, 'true': {'false': 423, 'true': 68105}}  
15                                               {'false': 0.375, 'true': 0.985}  
16                                                                          0.68  
17                                                                         0.966  
18                                               {'false': 0.985, 'true': 0.015}  
19                                                                           0.5  
20                                                                         0.046  
21                                               {'false': 0.006, 'true': 0.726}  
22                                                                         0.366  
23                                                                         0.703  
24                                               {'false': 0.015, 'true': 0.985}  
25                                                                           0.5  
26                                                                         0.954  
27                                               {'false': 0.426, 'true': 0.996}  
28                                                                         0.711  
29                                                                         0.977  
30                                               {'false': 0.593, 'true': 0.977}  
31                                                                         0.785  
32                                                                         0.964  
33                

In [115]:
params_dict = {'param': [],'value':[]}
v3_params_df = pd.DataFrame(params_dict, columns = ['param','value' ])

for param in v3_scores_model_info[project]['models'][model]['params'].keys():
    v3_params_df = v3_params_df.append({ 'param': param, 
                                        'value': v3_scores_model_info[project]['models'][model]['params'][param]}, 
                                         ignore_index=True)

In [116]:
print('Here is an overview about the parameters of the model \"reverted\" for the wikipedia project \"hrwiki\":')
v3_params_df

Here is an overview about the parameters of the model "reverted" for the wikipedia project "hrwiki":


param          value
0                  ccp_alpha              0
1                     center              1
2                  criterion   friedman_mse
3                       init           None
4              label_weights  {'false': 10}
5                     labels  [True, False]
6              learning_rate           0.01
7                       loss       deviance
8                  max_depth              3
9               max_features           log2
10            max_leaf_nodes           None
11     min_impurity_decrease              0
12        min_impurity_split           None
13          min_samples_leaf              5
14         min_samples_split              2
15  min_weight_fraction_leaf              0
16                multilabel          False
17              n_estimators            500
18          n_iter_no_change           None
19          population_rates           None
20                   presort     deprecated
21              random_state           None
22                     scale           True
23                 subsample              1
24                       tol         0.0001
25       validation_fraction            0.1
26                   verbose              0
27                warm_start          False

In [117]:
v3_scores_model_info[project]['models'][model]['environment'].keys()

dict_keys(['machine', 'platform', 'processor', 'python_branch', 'python_build', 'python_compiler', 'python_implementation', 'python_revision', 'python_version', 'release', 'revscoring_version', 'system', 'version'])

In [118]:
environment_dict = {'environment property': [],'value':[]}
v3_environment_df = pd.DataFrame(environment_dict, columns = ['environment property','value' ])

for env_prop in v3_scores_model_info[project]['models'][model]['environment'].keys():
    v3_environment_df = v3_environment_df.append({ 'environment property': env_prop, 
                                        'value': v3_scores_model_info[project]['models'][model]['environment'][env_prop]}, 
                                         ignore_index=True)

In [119]:
print('Here is an overview about the environment properties of the model \"reverted\" for the wikipedia project \"hrwiki\":')
v3_environment_df

Here is an overview about the environment properties of the model "reverted" for the wikipedia project "hrwiki":


environment property                                         value
0                 machine                                        x86_64
1                platform  Linux-4.9.0-11-amd64-x86_64-with-debian-9.12
2               processor                                              
3           python_branch                                              
4            python_build               [default, Sep 27 2018 17:25:39]
5         python_compiler                            GCC 6.3.0 20170516
6   python_implementation                                       CPython
7         python_revision                                              
8          python_version                                         3.5.3
9                 release                                4.9.0-11-amd64
10     revscoring_version                                         2.8.0
11                 system                                         Linux
12                version   #1 SMP Debian 4.9.189-3+deb9u1 (2019-09-20)

> The API call `https://ores.wikimedia.org/v3/scores/hrwiki` provides information about available model/ version iformation for a wikipedia project of interest. 

________________________________________________

In [120]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki')
v3_scores_hrwiki = api_call.json()

In [121]:
print('For the hrwiki project only the model \"reverted\" is available in version 0.5.0.')
pd.DataFrame.from_dict(v3_scores_hrwiki['hrwiki']['models'])

For the hrwiki project only the model "reverted" is available in version 0.5.0.


reverted
version    0.5.0

The API call `https://ores.wikimedia.org/v3/scores/enwiki?models=reverted&model_info` provides the same information as the call `https://ores.wikimedia.org/v3/scores/?model_info`. 
But with this API call the information can be requested **for the specific project and model of interest** where with the other API call information is returned for all projects and models. 

If the asked model is not available for the project then an error is returned: 

'code': 'not found',
 'message': "Models ('reverted',) not available for enwiki

________________________________________________

In [122]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki?models=reverted&model_info')
v3_scores_reverted_modelinfo = api_call.json()

In [123]:
v3_scores_reverted_modelinfo['hrwiki']['models']['reverted'].keys()

dict_keys(['environment', 'params', 'score_schema', 'statistics', 'type', 'version'])

In [124]:
stats_dict = {'metrics': [],'value':[]}
v3_stats_df = pd.DataFrame(stats_dict, columns = ['metrics','value' ])

for stat in v3_scores_reverted_modelinfo['hrwiki']['models']['reverted']['statistics'].keys():
    for label in v3_scores_reverted_modelinfo['hrwiki']['models']['reverted']['statistics'][stat].keys():
        v3_stats_df = v3_stats_df.append({ 'metrics':stat+' ('+label+')', 
                                          'value': v3_scores_reverted_modelinfo['hrwiki']['models']['reverted']['statistics'][stat][label]}, 
                                         ignore_index=True)

In [125]:
print('We checked if the stats are realy the same as the general modelinfo API call provides for the reverted model. We can confirm that it does provide consistent information.')
v3_stats_df

We checked if the stats are realy the same as the general modelinfo API call provides for the reverted model. We can confirm that it does provide consistent information.


metrics  \
0           !f1 (labels)   
1            !f1 (macro)   
2            !f1 (micro)   
3    !precision (labels)   
4     !precision (macro)   
5     !precision (micro)   
6       !recall (labels)   
7        !recall (macro)   
8        !recall (micro)   
9      accuracy (labels)   
10      accuracy (macro)   
11      accuracy (micro)   
12       counts (labels)   
13            counts (n)   
14  counts (predictions)   
15           f1 (labels)   
16            f1 (macro)   
17            f1 (micro)   
18  filter_rate (labels)   
19   filter_rate (macro)   
20   filter_rate (micro)   
21          fpr (labels)   
22           fpr (macro)   
23           fpr (micro)   
24   match_rate (labels)   
25    match_rate (macro)   
26    match_rate (micro)   
27       pr_auc (labels)   
28        pr_auc (macro)   
29        pr_auc (micro)   
30    precision (labels)   
31     precision (macro)   
32     precision (micro)   
33    rates (population)   
34        rates (sample)   
35       recall (labels)   
36        recall (macro)   
37        recall (micro)   
38      roc_auc (labels)   
39       roc_auc (macro)   
40       roc_auc (micro)   

                                                                              value  
0                                                   {'false': 0.494, 'true': 0.919}  
1                                                                             0.707  
2                                                                             0.527  
3                                                   {'false': 0.347, 'true': 0.986}  
4                                                                             0.666  
5                                                                             0.398  
6                                                   {'false': 0.855, 'true': 0.862}  
7                                                                             0.858  
8                                                                             0.855  
9                                                   {'false': 0.861, 'true': 0.861}  
10                                                                            0.861  
11                                                                            0.861  
12                                                   {'false': 18232, 'true': 1452}  
13                                                                            19684  
14  {'false': {'false': 15708, 'true': 2524}, 'true': {'false': 211, 'true': 1241}}  
15                                                  {'false': 0.919, 'true': 0.494}  
16                                                                            0.707  
17                                                                            0.886  
18                                                  {'false': 0.195, 'true': 0.805}  
19                                                                              0.5  
20                                                                            0.244  
21                                                  {'false': 0.145, 'true': 0.138}  
22                                                                            0.142  
23                                                                            0.145  
24                                                  {'false': 0.805, 'true': 0.195}  
25                                                                              0.5  
26                                                                            0.756  
27                                                  {'false': 0.992, 'true': 0.527}  
28                                                                             0.76  
29                                                                            0.956  
30                                                  {'false': 0.986, 'true': 0.347}  
31                                                                            0.666  
32

> The API call `https://ores.wikimedia.org/v3/scores/hrwiki?models=reverted&revids=235107991` returns a prediction for a specific wikipedia article. From its output the following ifnormation can be captured: 
<br>
<br>A **prediction** stating if the revision problbly needs to be reverted or not.
<br>A **probability for false** that the revision needs to be reverted.
<br>A **probability for true** that the revision needs to be reverted. 

________________________________________________

In [126]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki?models=reverted&revids=5618117')
v3_scores_reverted_revid = api_call.json()

In [127]:
print('The score schema (from the modelinfo API call) of the model \"reverted\" in the project \"hrwiki\":')
v3_scores_model_info['hrwiki']['models']['reverted']['score_schema']

The score schema (from the modelinfo API call) of the model "reverted" in the project "hrwiki":


{'properties': {'prediction': {'description': 'The most likely label predicted by the estimator',
   'type': 'boolean'},
  'probability': {'description': 'A mapping of probabilities onto each of the potential output labels',
   'properties': {'false': {'type': 'number'}, 'true': {'type': 'number'}},
   'type': 'object'}},
 'title': 'Scikit learn-based classifier score with probability',
 'type': 'object'}

In [128]:
print('This is how a prediction of the reverted model looks like for a article revision which does not need to be reverted:')
v3_scores_reverted_revid['hrwiki']['scores']

This is how a prediction of the reverted model looks like for a article revision which does not need to be reverted:


{'5618117': {'reverted': {'score': {'prediction': False,
    'probability': {'false': 0.8265831588525849,
     'true': 0.17341684114741504}}}}}

> The API call `https://ores.wikimedia.org/v3/scores/elwiki/807457197/reverted?features=true` returns information about the models features it is looking at for making a prediction about an article revision.

In [129]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki/5618117/reverted?features=true')
v3_scores_reverted_features_revid = api_call.json()

In [130]:
v3_scores_reverted_features_revid['hrwiki']['scores']['5618117']['reverted'].keys()

dict_keys(['features', 'score'])

In [131]:
feature_dict = {'feature': [],'value':[]}
v3_features_df = pd.DataFrame(feature_dict, columns = ['feature','value' ])

for feature in v3_scores_reverted_features_revid['hrwiki']['scores']['5618117']['reverted']['features'].keys():
    v3_features_df = v3_features_df.append({ 'feature': feature, 'value':v3_scores_reverted_features_revid['hrwiki']['scores']['5618117']['reverted']['features'][feature] }, ignore_index=True)

In [132]:
print('The reverted model has ',len(v3_features_df),' features. But most of them have the vlaue 0.')
v3_features_df


The reverted model has  78  features. But most of them have the vlaue 0.


feature  \
0           feature.croatian.badwords.revision.diff.match_delta_decrease   
1           feature.croatian.badwords.revision.diff.match_delta_increase   
2                feature.croatian.badwords.revision.diff.match_delta_sum   
3      feature.croatian.badwords.revision.diff.match_prop_delta_decrease   
4      feature.croatian.badwords.revision.diff.match_prop_delta_increase   
5           feature.croatian.badwords.revision.diff.match_prop_delta_sum   
6          feature.croatian.informals.revision.diff.match_delta_decrease   
7          feature.croatian.informals.revision.diff.match_delta_increase   
8               feature.croatian.informals.revision.diff.match_delta_sum   
9     feature.croatian.informals.revision.diff.match_prop_delta_decrease   
10    feature.croatian.informals.revision.diff.match_prop_delta_increase   
11         feature.croatian.informals.revision.diff.match_prop_delta_sum   
12           feature.english.badwords.revision.diff.match_delta_decrease   
13           feature.english.badwords.revision.diff.match_delta_increase   
14                feature.english.badwords.revision.diff.match_delta_sum   
15      feature.english.badwords.revision.diff.match_prop_delta_decrease   
16      feature.english.badwords.revision.diff.match_prop_delta_increase   
17           feature.english.badwords.revision.diff.match_prop_delta_sum   
18          feature.english.informals.revision.diff.match_delta_decrease   
19          feature.english.informals.revision.diff.match_delta_increase   
20               feature.english.informals.revision.diff.match_delta_sum   
21     feature.english.informals.revision.diff.match_prop_delta_decrease   
22     feature.english.informals.revision.diff.match_prop_delta_increase   
23          feature.english.informals.revision.diff.match_prop_delta_sum   
24  feature.len(<datasource.tokenized(datasource.revision.parent.text)>)   
25         feature.len(<datasource.tokenized(datasource.revision.text)>)   
26                   feature.len(<datasource.wikitext.revision.markups>)   
27            feature.len(<datasource.wikitext.revision.parent.markups>)   
28    feature.len(<datasource.wikitext.revision.parent.uppercase_words>)   
29              feature.len(<datasource.wikitext.revision.parent.words>)   
30                     feature.len(<datasource.wikitext.revision.words>)   
31                                     feature.revision.comment.has_link   
32                        feature.revision.comment.suggests_section_edit   
33                       feature.revision.diff.longest_new_repeated_char   
34                               feature.revision.diff.longest_new_token   
35                                   feature.revision.page.is_articleish   
36                                   feature.revision.page.is_draftspace   
37                                    feature.revision.page.is_mainspace   
38                             feature.revision.user.has_advanced_rights   
39                                        feature.revision.user.is_admin   
40                                         feature.revision.user.is_anon   
41                                          feature.revision.user.is_bot   
42                                      feature.revision.user.is_curator   
43                                    feature.revision.user.is_patroller   
44                                      feature.revision.user.is_trusted   
45             feature.temporal.revision.user.seconds_since_registration   
46                                       feature.wikitext.revision.chars   
47                  feature.wikitext.revision.diff.markup_delta_decrease   
48                  feature.wikitext.revision.diff.markup_delta_increase   
49                       feature.wikitext.revision.diff.markup_delta_sum   
50             feature.wikitext.revision.diff.markup_prop_delta_decrease   
51             feature.wikitext.revision.diff.markup_prop_delta_increase   
52                  feature.wikitext.

In [133]:
print('Number of features where the value is different from 0: ',len(v3_features_df[v3_features_df['value'] != 0]))
v3_features_df[v3_features_df['value'] != 0]
df = v3_features_df[v3_features_df['value'] != 0]
print(tabulate(df, tablefmt="pipe", headers="keys"))

Number of features where the value is different from 0:  29
|    | feature                                                              |           value |
|---:|:---------------------------------------------------------------------|----------------:|
| 24 | feature.len(<datasource.tokenized(datasource.revision.parent.text)>) |  9877           |
| 25 | feature.len(<datasource.tokenized(datasource.revision.text)>)        |  9902           |
| 26 | feature.len(<datasource.wikitext.revision.markups>)                  |  3384           |
| 27 | feature.len(<datasource.wikitext.revision.parent.markups>)           |  3380           |
| 28 | feature.len(<datasource.wikitext.revision.parent.uppercase_words>)   |    46           |
| 29 | feature.len(<datasource.wikitext.revision.parent.words>)             |  2083           |
| 30 | feature.len(<datasource.wikitext.revision.words>)                    |  2089           |
| 32 | feature.revision.comment.suggests_section_edit                       

> The API call `https://ores.wikimedia.org/v3/scores/enwiki/REVID/YOURMODELNAME?model_info` returns information about the model and the score for the spcified article revision. I think it is especialy usefull when checking some performance metrices like the *precision* or *recall* is important along with the with the *score*.  

________________________________________________

In [134]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki/5618117/reverted?model_info')
v3_scores_modelinfo_revision = api_call.json()

In [135]:
print("The modelinfo API call provides information on following attributes of all models:")
print(pd.DataFrame(v3_scores_modelinfo_revision['hrwiki']['models']['reverted'].keys()))

The modelinfo API call provides information on following attributes of all models:
              0
0   environment
1        params
2  score_schema
3    statistics
4          type
5       version


In [136]:
stats_dict = {'metrics': [],'value':[]}
v3_stats_df = pd.DataFrame(stats_dict, columns = ['metrics','value' ])

for stat in v3_scores_modelinfo_revision['hrwiki']['models']['reverted']['statistics'].keys():
    for label in v3_scores_modelinfo_revision['hrwiki']['models']['reverted']['statistics'][stat].keys():
        v3_stats_df = v3_stats_df.append({ 'metrics':stat+' ('+label+')', 
                                          'value': v3_scores_modelinfo_revision['hrwiki']['models']['reverted']['statistics'][stat][label]}, 
                                         ignore_index=True)

In [137]:
v3_stats_df

metrics  \
0           !f1 (labels)   
1            !f1 (macro)   
2            !f1 (micro)   
3    !precision (labels)   
4     !precision (macro)   
5     !precision (micro)   
6       !recall (labels)   
7        !recall (macro)   
8        !recall (micro)   
9      accuracy (labels)   
10      accuracy (macro)   
11      accuracy (micro)   
12       counts (labels)   
13            counts (n)   
14  counts (predictions)   
15           f1 (labels)   
16            f1 (macro)   
17            f1 (micro)   
18  filter_rate (labels)   
19   filter_rate (macro)   
20   filter_rate (micro)   
21          fpr (labels)   
22           fpr (macro)   
23           fpr (micro)   
24   match_rate (labels)   
25    match_rate (macro)   
26    match_rate (micro)   
27       pr_auc (labels)   
28        pr_auc (macro)   
29        pr_auc (micro)   
30    precision (labels)   
31     precision (macro)   
32     precision (micro)   
33    rates (population)   
34        rates (sample)   
35       recall (labels)   
36        recall (macro)   
37        recall (micro)   
38      roc_auc (labels)   
39       roc_auc (macro)   
40       roc_auc (micro)   

                                                                              value  
0                                                   {'false': 0.494, 'true': 0.919}  
1                                                                             0.707  
2                                                                             0.527  
3                                                   {'false': 0.347, 'true': 0.986}  
4                                                                             0.666  
5                                                                             0.398  
6                                                   {'false': 0.855, 'true': 0.862}  
7                                                                             0.858  
8                                                                             0.855  
9                                                   {'false': 0.861, 'true': 0.861}  
10                                                                            0.861  
11                                                                            0.861  
12                                                   {'false': 18232, 'true': 1452}  
13                                                                            19684  
14  {'false': {'false': 15708, 'true': 2524}, 'true': {'false': 211, 'true': 1241}}  
15                                                  {'false': 0.919, 'true': 0.494}  
16                                                                            0.707  
17                                                                            0.886  
18                                                  {'false': 0.195, 'true': 0.805}  
19                                                                              0.5  
20                                                                            0.244  
21                                                  {'false': 0.145, 'true': 0.138}  
22                                                                            0.142  
23                                                                            0.145  
24                                                  {'false': 0.805, 'true': 0.195}  
25                                                                              0.5  
26                                                                            0.756  
27                                                  {'false': 0.992, 'true': 0.527}  
28                                                                             0.76  
29                                                                            0.956  
30                                                  {'false': 0.986, 'true': 0.347}  
31                                                                            0.666  
32

#### Conclusion:


### Feature Injection
Please check out the _feature injection_ feature of ORES: https://www.mediawiki.org/wiki/ORES/Feature_injection

**Example:**

     # Here you can get the perdiction for a revision, if the user would habe been anonymous:
     https://ores.wikimedia.org/v3/scores/enwiki/991397091/damaging?features&feature.revision.user.is_anon=true

Some of the features we think could be intresting to look at: 
    
   **User** related freatures:
    
    feature.revision.user.has_advanced_rights
    feature.revision.user.is_admin
    feature.revision.user.is_anon
    feature.revision.user.is_bot
    feature.revision.user.is_curator
    feature.revision.user.is_patroller
    feature.revision.user.is_trusted
    feature.temporal.revision.user.seconds_since_registration
    feature.enwiki.revision.cite_templates

   **Bad words** and language related:
    
    feature.croatian.badwords.revision.diff.match_delta_decrease
    feature.croatian.badwords.revision.diff.match_delta_increase
    feature.croatian.badwords.revision.diff.match_delta_sum	
    feature.croatian.badwords.revision.diff.match_prop_delta_decrease
    feature.croatian.badwords.revision.diff.match_prop_delta_increase
    feature.croatian.badwords.revision.diff.match_prop_delta_sum
    
    feature.english.badwords.revision.diff.match_delta_decrease	
    feature.english.badwords.revision.diff.match_delta_increase	
    feature.english.badwords.revision.diff.match_delta_sum	
    feature.english.badwords.revision.diff.match_prop_delta_decrease
    feature.english.badwords.revision.diff.match_prop_delta_increase
    feature.english.badwords.revision.diff.match_prop_delta_sum	


In [138]:
feature_injection_dict = {'injected feature': [], 'probability (false)':[], 'probability (true)':[],  'prediction':[]}
feature_injection_df = pd.DataFrame(feature_injection_dict, columns = ['injected feature','probability (false)','probability (true)','prediction'  ])

#### Without feature injection

In [139]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki?models=reverted&revids=5618117')
reverted_score_wo_injection = api_call.json()

In [140]:
df = reverted_score_wo_injection 
feature_injection_df = feature_injection_df.append({'injected feature': 'without feature injection', 
                                               'probability (false)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['false'],  
                                               'probability (true)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['true'],
                                               'prediction':df['hrwiki']['scores']['5618117']['reverted']['score']['prediction']},
                                                ignore_index=True)

In [141]:
feature_injection_df

injected feature  probability (false)  probability (true)  \
0  without feature injection             0.826583            0.173417   

   prediction  
0         0.0

In [142]:
reverted_score_wo_injection['hrwiki']['scores']['5618117']['reverted']['score']['probability']['false']

0.8265831588525849

#### With injecting the feature `user.is_bot=true`

In [143]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki/5618117/reverted?features&feature.revision.user.is_bot=true')
reverted_score_user_bot = api_call.json()

In [144]:
df = reverted_score_user_bot
feature_injection_df = feature_injection_df.append({'injected feature': 'user.is_bot=true', 
                                               'probability (false)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['false'],  
                                               'probability (true)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['true'],
                                               'prediction':df['hrwiki']['scores']['5618117']['reverted']['score']['prediction']},
                                                ignore_index=True)

In [145]:
reverted_score_user_bot['hrwiki']['scores']['5618117']['reverted']['score']

{'prediction': False,
 'probability': {'false': 0.9585143260758429, 'true': 0.04148567392415714}}

This was unexpected: The probability the treu probability for this revision to be rejected decreases if the user is a bot.

#### With injecting the feature `user.is_trusted=false`

In [146]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki/5618117/reverted?features&feature.revision.user.is_trusted=false')
reverted_score_user_trusted = api_call.json()

In [147]:
df = reverted_score_user_trusted
feature_injection_df = feature_injection_df.append({'injected feature': 'is_trusted=false', 
                                               'probability (false)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['false'],  
                                               'probability (true)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['true'],
                                               'prediction':df['hrwiki']['scores']['5618117']['reverted']['score']['prediction']},
                                                ignore_index=True)

In [148]:
reverted_score_user_trusted['hrwiki']['scores']['5618117']['reverted']['score']

{'prediction': False,
 'probability': {'false': 0.8265831588525849, 'true': 0.17341684114741504}}

This feature we testd with a numeric value as the modelinfo statistics is showing a '0.0' for this freature as value. But we also tried it with a boolean. 
But we noticed that both does not show a diffrence in the outcome compared to asking without feature injection.

#### With injecting the feature `feature.croatian.badwords.revision.diff.match_delta_increase`

In [149]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki/5618117/reverted?features&feature.croatian.badwords.revision.diff.match_delta_increase=2')
reverted_score_cratian_bad_word_increase = api_call.json()

In [150]:
df = reverted_score_cratian_bad_word_increase
feature_injection_df = feature_injection_df.append({'injected feature': 'feature.croatian.badwords.revision.diff.match_delta_increase=2', 
                                               'probability (false)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['false'],  
                                               'probability (true)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['true'],
                                               'prediction':df['hrwiki']['scores']['5618117']['reverted']['score']['prediction']},
                                                ignore_index=True)

In [151]:
reverted_score_cratian_bad_word_increase['hrwiki']['scores']['5618117']['reverted']['score']

{'prediction': False,
 'probability': {'false': 0.8125402216241415, 'true': 0.1874597783758585}}

This feature seems to be important as it also changes the probability of this revision into the negative direction.

#### With injecting the feature `feature.english.badwords.revision.diff.match_delta_increase`

In [152]:
api_call = requests.get('https://ores.wikimedia.org/v3/scores/hrwiki/5618117/reverted?features&feature.english.badwords.revision.diff.match_delta_increase=2')
reverted_score_english_bad_word_increase = api_call.json()

In [153]:
df = reverted_score_english_bad_word_increase
feature_injection_df = feature_injection_df.append({'injected feature': 'feature.english.badwords.revision.diff.match_delta_increase=2', 
                                               'probability (false)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['false'],  
                                               'probability (true)':df['hrwiki']['scores']['5618117']['reverted']['score']['probability']['true'],
                                               'prediction':df['hrwiki']['scores']['5618117']['reverted']['score']['prediction']},
                                                ignore_index=True)

In [154]:
reverted_score_english_bad_word_increase['hrwiki']['scores']['5618117']['reverted']['score']

{'prediction': False,
 'probability': {'false': 0.8265831588525849, 'true': 0.17341684114741504}}

But an increase of bad words count in english language does not change the score. 

In [155]:
feature_injection_df

injected feature  \
0                                       without feature injection   
1                                                user.is_bot=true   
2                                                is_trusted=false   
3  feature.croatian.badwords.revision.diff.match_delta_increase=2   
4   feature.english.badwords.revision.diff.match_delta_increase=2   

   probability (false)  probability (true)  prediction  
0             0.826583            0.173417         0.0  
1             0.958514            0.041486         0.0  
2             0.826583            0.173417         0.0  
3             0.812540            0.187460         0.0  
4             0.826583            0.173417         0.0

|    | injected feature                                               |   probability (false) |   probability (true) |   prediction |
|---:|:---------------------------------------------------------------|----------------------:|---------------------:|-------------:|
|  0 | without feature injection                                      |              0.826583 |            0.173417  |            0 |
|  1 | user.is_bot=true                                               |              0.958514 |            0.0414857 |            0 |
|  2 | is_trusted=false                                               |              0.826583 |            0.173417  |            0 |
|  3 | feature.croatian.badwords.revision.diff.match_delta_increase=2 |              0.81254  |            0.18746   |            0 |
|  4 | feature.english.badwords.revision.diff.match_delta_increase=2  |              0.826583 |            0.173417  |            0 |

### Conclusion:

## [3] ML algorithm and training/test data
> Which machine learning model is underlying and what data is used to build the model?

* Check out `model_info` in detail.
* What does it tell you about the model performance?
* You can visualise and explain your results regarding model performance.
* What data was used to train and test the model?
* What machine learning algorithm is your model using? Please explain briefly.

* Check out `model_info` in detail.


* What does it tell you about the model performance?

`model_info` provides several measures of the model's performance.
Under the attribute `statistics` you can find the values for the confusion matrix, several related measures (recall, precision, false-positive rate) as well as AUC (Area Under Curve) for the ROC curve (Receiver Operating Characteristics).

In [156]:
v3_stats_df[v3_stats_df['metrics'].str.startswith('precision') | 
            v3_stats_df['metrics'].str.startswith('recall') | 
            v3_stats_df['metrics'].str.startswith('f1')]

metrics                            value
15         f1 (labels)  {'false': 0.919, 'true': 0.494}
16          f1 (macro)                            0.707
17          f1 (micro)                            0.886
30  precision (labels)  {'false': 0.986, 'true': 0.347}
31   precision (macro)                            0.666
32   precision (micro)                            0.935
35     recall (labels)  {'false': 0.862, 'true': 0.855}
36      recall (macro)                            0.858
37      recall (micro)                            0.861

In [157]:
v3_stats_df[v3_stats_df['metrics'] == 'counts (predictions)']

metrics  \
14  counts (predictions)   

                                                                              value  
14  {'false': {'false': 15708, 'true': 2524}, 'true': {'false': 211, 'true': 1241}}

* Precision (TP) / (TP + FP):

Precision tells us how relevant are the positive detections. Higher the precision better is our detection mechanism. e.g. in our example precision is just 10% which is very poor. We are misclassifying a lot of healthy patients as cancer patients.


* Recall (sensitivity/ TPR) (TP)/(TP + FN):
What percentage of actual positives are detected. i.e. what percentage of real cancer patients are detected. It is computed by taking the ratio of correctly identified cancer patients (true positives) to the total number of cancer patients (true positives + false negatives).



In [158]:
predictions_dict = {'false': [15708, 2524], 'true': [ 211, 1241]}
confusion_matrix = pd.DataFrame(predictions_dict, columns = ['true','false'], index=['false','true'])

In [159]:
print('Confusion matrix: ')
confusion_matrix 

Confusion matrix: 


true  false
false   211  15708
true   1241   2524

* What data was used to train and test the model?

The history of edits (and reverted edits) from a wiki.

* What machine learning algorithm is your model using? Please explain briefly. \
Different algorithms were used. Most wikis use GradientBoosting (bnwiki, elwiki, glwiki, hrwiki, idwiki, iswiki, tawiki, viwiki). Gradient boosting creates an ensemble learner by iteratively adding weak learners to an ensemble. Only the enwiktionary uses RandomForest and testwiki uses RevIDScorer.

## [4] Features
> Which features are used and which have the greatest influence on the prediction?

* What features is your model using?


* What do they mean?
* Which is the most important features?
* `https://ores.wikimedia.org/v3/scores/enwiki/991379667/articlequality?features=true`
* Are all models (in all languages of wikipedia), are they using the same features?

## Sample code

In [160]:


# Customize these with your own information
headers = {
    'User-Agent': 'https://github.com/YOUR-USER-NAME',
    'From': 'YOUR-EMAIL@fu-berlin.de'
}

def get_ores_data(rev_id, headers):
    
    # Define the endpoint: This is an example!
    endpoint = 'https://ores.wikimedia.org/v3/scores/{project}/?models={model}&revids={revids}'

    params = {'project' : 'enwiki',
              'model'   : 'YOUMODELNAME',
              'revids'  : rev_id
              }

    api_call = requests.get(endpoint.format(**params))
    response = api_call.json()
    data = json.dumps(response)

    return data

***

#### Credits

We release the notebooks under the [Creative Commons Attribution license (CC BY 4.0)](https://creativecommons.org/licenses/by/4.0/).